// AST/index.js
"use strict";
var Data_List_Types = require("../Data.List.Types/index.js");

// and for now Ethereals are just polarGridHelpers...
var PolarGridHelper = (function () {
    function PolarGridHelper(value0) {
        this.value0 = value0;
    };
    PolarGridHelper.create = function (value0) {
        return new PolarGridHelper(value0);
    };
    return PolarGridHelper;
})();

// Element-s can be Dancer-s (which can be animated, have physics, etc)
// or Ethereal-s (which are displayed but do not have physics [maybe they might have animation, though?])
var Dancer = (function () {
    function Dancer(value0) {
        this.value0 = value0;
    };
    Dancer.create = function (value0) {
        return new Dancer(value0);
    };
    return Dancer;
})();

// Element-s can be Dancer-s (which can be animated, have physics, etc)
// or Ethereal-s (which are displayed but do not have physics [maybe they might have animation, though?])
var Ethereal = (function () {
    function Ethereal(value0) {
        this.value0 = value0;
    };
    Ethereal.create = function (value0) {
        return new Ethereal(value0);
    };
    return Ethereal;
})();

// Statement-s describe an Element in the scene
// or they change some other parameter of the rendering (eg. camera settings)
var Element = (function () {
    function Element(value0) {
        this.value0 = value0;
    };
    Element.create = function (value0) {
        return new Element(value0);
    };
    return Element;
})();

// Statement-s describe an Element in the scene
// or they change some other parameter of the rendering (eg. camera settings)
var CameraChange = (function () {
    function CameraChange(value0) {
        this.value0 = value0;
    };
    CameraChange.create = function (value0) {
        return new CameraChange(value0);
    };
    return CameraChange;
})();
var origin = {
    x: 0.0,
    y: 0.0,
    z: 0.0
};
var defaultProgram = Data_List_Types.Nil.value;
var defaultEthereal = new PolarGridHelper({
    radius: 10.0,
    radials: 16,
    circles: 8,
    divisions: 64,
    position: origin
});
var defaultDancer = {
    url: "",
    animation: 4,
    position: origin
};
module.exports = {
    defaultProgram: defaultProgram,
    Element: Element,
    CameraChange: CameraChange,
    Dancer: Dancer,
    Ethereal: Ethereal,
    defaultDancer: defaultDancer,
    PolarGridHelper: PolarGridHelper,
    defaultEthereal: defaultEthereal
};
