// Halogen.HTML.Elements.Keyed/index.js
"use strict";
var Halogen_HTML_Elements = require("../Halogen.HTML.Elements/index.js");
var ul = Halogen_HTML_Elements.keyed("ul");
var ul_ = ul([  ]);
var tr = Halogen_HTML_Elements.keyed("tr");
var tr_ = tr([  ]);
var thead = Halogen_HTML_Elements.keyed("thead");
var thead_ = thead([  ]);
var tfoot = Halogen_HTML_Elements.keyed("tfoot");
var tfoot_ = tfoot([  ]);
var tbody = Halogen_HTML_Elements.keyed("tbody");
var tbody_ = tbody([  ]);
var table = Halogen_HTML_Elements.keyed("table");
var table_ = table([  ]);
var ol = Halogen_HTML_Elements.keyed("ol");
var ol_ = ol([  ]);
var menu = Halogen_HTML_Elements.keyed("menu");
var menu_ = menu([  ]);
var header = Halogen_HTML_Elements.keyed("header");
var header_ = header([  ]);
var form = Halogen_HTML_Elements.keyed("form");
var form_ = form([  ]);
var footer = Halogen_HTML_Elements.keyed("footer");
var footer_ = footer([  ]);
var fieldset = Halogen_HTML_Elements.keyed("fieldset");
var fieldset_ = fieldset([  ]);
var dl = Halogen_HTML_Elements.keyed("dl");
var dl_ = dl([  ]);
var div = Halogen_HTML_Elements.keyed("div");
var div_ = div([  ]);
var dialog = Halogen_HTML_Elements.keyed("dialog");
var dialog_ = dialog([  ]);
var colgroup = Halogen_HTML_Elements.keyed("colgroup");
var colgroup_ = colgroup([  ]);
var article = Halogen_HTML_Elements.keyed("article");
var article_ = article([  ]);
module.exports = {
    article: article,
    article_: article_,
    colgroup: colgroup,
    colgroup_: colgroup_,
    dialog: dialog,
    dialog_: dialog_,
    div: div,
    div_: div_,
    dl: dl,
    dl_: dl_,
    fieldset: fieldset,
    fieldset_: fieldset_,
    footer: footer,
    footer_: footer_,
    form: form,
    form_: form_,
    header: header,
    header_: header_,
    menu: menu,
    menu_: menu_,
    ol: ol,
    ol_: ol_,
    table: table,
    table_: table_,
    tbody: tbody,
    tbody_: tbody_,
    tfoot: tfoot,
    tfoot_: tfoot_,
    thead: thead,
    thead_: thead_,
    tr: tr,
    tr_: tr_,
    ul: ul,
    ul_: ul_
};
