// Graphics.Three.Geometry/index.js
"use strict";
var Graphics_Three_Util = require("../Graphics.Three.Util/index.js");
var setVertices = Graphics_Three_Util.fpi([ "geometry", "vertices", "" ])("geometry.vertices = vertices");
var getVertices = Graphics_Three_Util.ffi([ "geometry", "" ])("geometry.vertices");
var createPlane = Graphics_Three_Util.ffi([ "width", "height", "widthSegment", "heightSegment", "" ])("new THREE.PlaneGeometry(width, height, widthSegment, heightSegment);");
var createCircle = Graphics_Three_Util.ffi([ "radius", "segments", "thetaStart", "thetaLength", "" ])("new THREE.CircleGeometry(radius, segments, thetaStart, thetaLength)");
var createBox = Graphics_Three_Util.ffi([ "width", "height", "depth", "" ])("new THREE.BoxGeometry(width, height, depth)");
var create = Graphics_Three_Util.ffi([ "vertices", "" ])(" (function() { \x0a            var geometry = new THREE.Geometry(); \x0a            geometry.vertices = vertices\x0a            return geometry;\x0a        }())\x0a    ");
var addVertex = Graphics_Three_Util.fpi([ "geometry", "vertex", "" ])("geometry.vertices.push(vertex)");
module.exports = {
    create: create,
    createBox: createBox,
    createCircle: createCircle,
    createPlane: createPlane,
    getVertices: getVertices,
    setVertices: setVertices,
    addVertex: addVertex
};
